### PR TITLE
add percent tumour cell exception validations

### PR DESCRIPTION
**Pull request by justincorrigible**

### CODE REVIEW

File: references/validationFunctions/specimen/percentageTumourCells.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    // checks for a string just consisting of whitespace
+    const checkforEmpty = (entry) => {
+      return /^\s+$/g.test(decodeURI(entry).replace(/^"(.*)"$/, '$1'));
+    };
+
+    const fieldHasValue = $field && $field != null && !(checkforEmpty($field));
+    const measurementMethod = $row?.percent_tumour_cells_measurement_method?.trim?.().toLowerCase();
+
+    if (fieldHasValue) {
+      if (measurementMethodExceptionTypes.includes(measurementMethod)) {
```
Comment (justincorrigible) [inline references/validationFunctions/specimen/percentageTumourCells.js:38]: if `percent_tumour_cell` has a value AND  `percent_tumour_cell_measurement_method` is "not applicable" (exceptions list is extendable here to "unknown" etc.), then validation fails

File: references/validationFunctions/specimen/percentageTumourCells.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        return {
+          valid: false,
+          message: `The '${$name}' field cannot be submitted when 'percent_tumour_cells_measurement_method' = 'Not applicable'`
+        };
+      }
+    }
+
+    return result;
```
Comment (justincorrigible) [inline references/validationFunctions/specimen/percentageTumourCells.js:46]: ...otherwise it passes

File: schemas/specimen.json [hunk 3]
```diff
@@ -349,9 +351,9 @@
       },
       "restrictions": {
         "codeList": [
-           "Genomics",
-           "Image analysis",
-           "Pathology estimate by percent nuclei"
+          "Genomics",
+          "Image analysis",
+          "Pathology estimate by percent nuclei"
```
Comment (justincorrigible) [inline schemas/specimen.json:356]: Came looking for these values, for the tests... ended up cleaning some whitespace 👍

File: tests/specimen/percentageTumourCells.test.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+const universalTest = require('../universal');
+const loadObjects = require('../loadObjects');
+
+// load in all fields with entries prepopulated to null
+const specimen = require('../constructDummyData').getSchemaDummy('specimen');
+
+// the name of the field being validateds
+const name = 'percent_tumour_cells';
+
+const unitTests = [
```
Comment (justincorrigible) [inline tests/specimen/percentageTumourCells.test.js:31]: A few combinations, which surely can be extended. Open to suggestions/corrections.

File: schemas/specimen.json [hunk 5]
```diff
@@ -331,10 +333,10 @@
...
       "restrictions": {
-         "range": {
-            "min": 0,
-            "max": 1
-          }
+        "range": {
+          "min": 0,
+          "max": 1
+        }
       }
```
Comment (joneubank) [inline schemas/specimen.json:340]: Pretty sure you meant to add the script here
```json
"script": "#/script/specimen/percentageTumourCells"
```

Comment (justincorrigible) [reply schemas/specimen.json:340]: Fairly sure you're right. The problem with rushing out the door lol

thanks!

**Review by joneubank — APPROVED** [review]

**Pull request merged**